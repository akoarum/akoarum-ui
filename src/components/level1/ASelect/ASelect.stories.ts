import { storiesOf } from '@storybook/vue'
import { withKnobs, text } from '@storybook/addon-knobs'
import { withInfo } from 'storybook-addon-vue-info'
import ASelect from './ASelect.vue'
import { action } from '@storybook/addon-actions'

storiesOf('level1/ASelect', module)
  .addDecorator(withInfo)
  .addDecorator(withKnobs)
  .add('Normal', () => ({
    components: { ASelect },
    props: {
      placeholder: { default: text('プレスホルダー', '選択してください') },
      width: { default: text('幅', 'auto') }
    },
    data() {
      return {
        value: '',
        items: [
          { id: 1, label: '項目1', value: 1 },
          { id: 2, label: '項目2', value: 2 },
          { id: 3, label: '項目3', value: 3 }
        ]
      }
    },
    template: `
      <ASelect
        v-model="value"
        :items="items"
        :placeholder="placeholder"
        :width="width"
        @focus="focus"
        @blur="blur"
      />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
  .add('Error', () => ({
    components: { ASelect },
    props: {
      placeholder: { default: text('プレスホルダー', '選択してください') },
      width: { default: text('幅', 'auto') }
    },
    data() {
      return {
        value: '',
        items: [
          { id: 1, label: '項目1', value: 1 },
          { id: 2, label: '項目2', value: 2 },
          { id: 3, label: '項目3', value: 3 }
        ]
      }
    },
    template: `
      <ASelect
        v-model="value"
        :items="items"
        :placeholder="placeholder"
        :width="width"
        aria-invalid="true"
        @focus="focus"
        @blur="blur"
      />
    `,
    methods: {
      focus: action('focus'),
      blur: action('blur')
    }
  }), { info: {} })
