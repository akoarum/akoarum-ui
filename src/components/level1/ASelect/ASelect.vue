<template>
  <div class="aSelect">
    <select
      v-model="valueModel"
      v-bind="$attrs"
      :style="styled"
      class="aSelect__input"
      @focus="handleFocus"
      @blur="handleBlur"
    >
      <option value="">{{ placeholder }}</option>
      <template v-for="item in items">
        <option :key="item.id" :value="item.value">{{ item.value }}</option>
      </template>
    </select>
  </div>
</template>

<script lang="ts">
import { PropType } from 'vue'
import { defineComponent, computed, SetupContext } from '@vue/composition-api'
import { FormOption } from '~/types/FormOption'

interface Props {
  items: FormOption[]
  value: string | number | null
  width?: string
}

export default defineComponent({
  model: {
    event: 'change'
  },
  props: {
    items: { type: Array as PropType<FormOption[]>, required: true },
    value: { required: true },
    placeholder: { type: String, default: '選択してください' },
    width: { type: String, default: 'auto' }
  },
  setup(props: Props, context: SetupContext) {
    const styled = computed<{ [key: string]: string }>(() => ({
      '--width': props.width
    }))
    const valueModel = computed({
      get: (): string | number => props.value,
      set: (newValue: string | number) => {
        context.emit('change', newValue)
      }
    })
    const handleFocus = () => {
      context.emit('focus')
    }
    const handleBlur = () => {
      context.emit('blur')
    }

    return { styled, valueModel, handleFocus, handleBlur }
  }
})
</script>

<style lang="scss" scoped>
.aSelect {
  position: relative;
  display: inline-block;
  --width: auto;

  &::after {
    position: absolute;
    right: 6px;
    top: 9px;
    content: 'keyboard_arrow_down';
    font-size: 20px;
    font-family: 'Material Icons';
  }
}

.aSelect__input {
  display: inline-block;
  box-sizing: border-box;
  border: 1px solid $color-border;
  box-shadow: none;
  padding: 8px 38px 9px 12px;
  width: var(--width);
  font-size: 16px;
  line-height: 1.3;
  outline: none;
  transition: border $duration $easing, background $duration $easing;
  cursor: pointer;

  &[aria-invalid='true'] {
    border-color: $color-caution;
    background: #fff7f7;
  }

  &:focus {
    border-color: $color-primary;
  }
}
</style>
