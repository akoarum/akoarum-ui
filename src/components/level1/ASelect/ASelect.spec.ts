import { shallowMount, createLocalVue } from '@vue/test-utils'
import VueCompositionApi from '@vue/composition-api'
import ASelect from './ASelect.vue'

describe('ASelect', () => {
  const localVue = createLocalVue()
  let wrapper

  localVue.use(VueCompositionApi)

  beforeEach(() => {
    wrapper = shallowMount(ASelect, {
      propsData: {
        value: '',
        items: [
          { id: 1, label: '項目1', value: 1 },
          { id: 2, label: '項目2', value: 2 },
          { id: 3, label: '項目3', value: 3 }
        ]
      },
      localVue
    })
  })

  describe('initialized', () => {
    it('mountable', () => {
      expect(wrapper.isVueInstance()).toBe(true)
    })

    it('snapshot', () => {
      expect(wrapper.element).toMatchSnapshot()
    })

    it('styled: propsで幅を指定できる', () => {
      wrapper.setProps({ width: '120px' })
      expect(wrapper.vm.styled).toEqual({
        '--width': '120px'
      })
    })

    it('handleFocus: フォーカスイベントを通知する', () => {
      wrapper.find('select').trigger('focus')
      expect(wrapper.emitted().focus).toBeTruthy()
    })

    it('handleBlur: フォーカスが外れたことを通知する', () => {
      wrapper.find('select').trigger('blur')
      expect(wrapper.emitted().blur).toBeTruthy()
    })
  })
})
