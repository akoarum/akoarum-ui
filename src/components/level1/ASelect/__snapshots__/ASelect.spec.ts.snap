// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ASelect initialized snapshot 1`] = `
<div
  class="aSelect"
>
  <select
    class="aSelect__input"
    style="--width: auto;"
  >
    <option
      value=""
    >
      選択してください
    </option>
     
    <option
      value="1"
    >
      1
    </option>
    <option
      value="2"
    >
      2
    </option>
    <option
      value="3"
    >
      3
    </option>
  </select>
</div>
`;
